src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientBuilderExtensions.cs

```diff
@@ -650,11 +645,99 @@ public static IHttpClientBuilder ConfigureAdditionalHttpMessageHandlers(this IHt
             return builder;
         }
 
+        public static IHttpClientBuilder AddAsKeyed(this IHttpClientBuilder builder, ServiceLifetime lifetime = ServiceLifetime.Scoped)
+        {
+            ThrowHelper.ThrowIfNull(builder);
+
+            string? name = builder.Name;
+            IServiceCollection services = builder.Services;
+            HttpClientMappingRegistry registry = GetMappingRegistry(services);
+
+            UpdateEmptyNameHttpClient(services, registry);
+
+            if (name == null)
+            {
+                registry.DefaultKeyedLifetime?.RemoveRegistration(services);
+
+                registry.DefaultKeyedLifetime = new HttpClientKeyedLifetime(lifetime);
+                registry.DefaultKeyedLifetime.AddRegistration(services);
+            }
+            else
+            {
+                if (registry.KeyedLifetimeMap.TryGetValue(name, out HttpClientKeyedLifetime? clientLifetime))
+                {
+                    clientLifetime.RemoveRegistration(services);
+                }
+
+                clientLifetime = new HttpClientKeyedLifetime(name, lifetime);
+                registry.KeyedLifetimeMap[name] = clientLifetime;
+                clientLifetime.AddRegistration(services);
+            }
+
+            return builder;
+        }
+
+        public static IHttpClientBuilder RemoveAsKeyed(this IHttpClientBuilder builder)
+        {
+            ThrowHelper.ThrowIfNull(builder);
+
+            string? name = builder.Name;
+            IServiceCollection services = builder.Services;
+            HttpClientMappingRegistry registry = GetMappingRegistry(services);
+
+            UpdateEmptyNameHttpClient(services, registry);
+
+            if (name == null)
+            {
+                registry.DefaultKeyedLifetime?.RemoveRegistration(services);
+                registry.DefaultKeyedLifetime = HttpClientKeyedLifetime.Disabled;
+            }
+            else
+            {
+                if (registry.KeyedLifetimeMap.TryGetValue(name, out HttpClientKeyedLifetime? clientLifetime))
+                {
+                    clientLifetime.RemoveRegistration(services);
+                }
+                registry.KeyedLifetimeMap[name] = HttpClientKeyedLifetime.Disabled;
+            }
+
+            return builder;
+        }
+
+        // workaround for https://github.com/dotnet/runtime/issues/102654
+        private static void UpdateEmptyNameHttpClient(IServiceCollection services, HttpClientMappingRegistry registry)
```

Comment:
NIT: Please consider more specific name. Something in line of `EnsureDefaultHttpClientIsKeyd`

Reply:
"DefaultHttpClient" is unfortunately even more confusing, because it clashes with a notation of "ConfigureHttpClientDefaults":
 - the one I modify here is the one with `name=string.Empty` (so respective key would be also `string.Empty`),
 - while in "ConfigureHttpClientDefaults", `name=null` and the key is `KeyedService.AnyKey`

From that point of view, "EmptyNameHttpClient" describes exactly what it is... But I'll think whether there's a way to make it less confusing.

---

To make things even worse, you actually _can_ go and configure this "empty name" client in a way what would be different from all other unconfigured clients 🙃

```c#
services.ConfigureHttpClientDefaults(b =>
    b.ConfigureHttpClient(c => c.BaseAddress = "https://defaults.example.com"));

services.AddHttpClient("foo", c => c.BaseAddress = "https://foo.example.com");
services.AddHttpClient("unconfigured");

// AddHttpClient with the same name called twice does _not_ overwrite previous configs, only appends
services.AddHttpClient("")
    .ConfigureHttpClient(c => c.BaseAddress = "https://empty-name.example.com");

// ...

// will have "defaults. ..." base address
factory.CreateClient("unconfigured");

// will have "defaults. ..." base address
factory.CreateClient("some-absent-name");

// will have "empty-name. ..." base address :'D
// it's "default" in terms of resolving without a name, but not "default" in terms of configuration
serviceProvider.GetRequiredService<HttpClient>();
```

Reply:
Isn't this a little confusing, intuitively people could consider emptyname  and non keyd  as same.

Reply:
I agree... but then again, if there's NO keyed services at all, "empty name" client _is_ (historically and unfortunately) a non-keyed client (= available by `GetRequiredService<HttpClient>()`)

but as soon as you DO introduce anything keyed-services related, then (as a workaround for the issue I linked) there won't be a non-keyed client in the container anymore, and the "empty name" client will get updated to be accessible from container only by `GetRequiredKeyedService<HttpClient>(string.Empty)`, and `GetRequiredService<HttpClient>()` will start to throw.

Reply:
also:
- from _Keyed DI_ perspective,
   - **non-keyed** service is the one with `key=null`;
   - and `key=""` is just an **ordinary key** with no special behavior.
   - And key **"applies to all"** is `key=AnyKey` (non-null)
- and from the _factory_ (and _named IOptions_) perspective,
   - `name=null` means **"applies to all"**,
   - and `name=""` means an instance that you can get without specifying a name (hence the "default", like a default value of a parameter `(string name = string.Empty)`)

🙃🥲🙈

Reply:
I understand. Anyway, whoever choose to register keyd service with string.Empty will well deserve the consequences :-)